File: 09_Linked Lists/03_Move the Smallest and largest to head and tail of list.c
// Return the head of updated list
struct Node* shiftSmallLarge(struct Node *head) {
    if(head == NULL || head->next == NULL)
        return head;
    struct Node *curr = head, *prev = NULL;
    struct Node *minNode = head, *maxNode = head;
    struct Node *minPrev = NULL, *maxPrev = NULL;
    // Find min and max nodes
    while(curr != NULL)
    {
        if(curr->data < minNode->data)
        {
            minNode = curr;
            minPrev = prev;
        }
        if(curr->data > maxNode->data)
        {
            maxNode = curr;
            maxPrev = prev;
        }
        prev = curr;
        curr = curr->next;
    }
    // If min and max are same
    if(minNode == maxNode)
        return head;
    // -------- Move minimum to head --------
    if(minNode != head)
    {
        minPrev->next = minNode->next;
        minNode->next = head;
        head = minNode;
    }
    // Recalculate maxPrev if maxNode was before minNode
    prev = NULL;
    curr = head;
    while(curr != NULL && curr != maxNode)
    {
        prev = curr;
        curr = curr->next;
    }
    maxPrev = prev;
    // -------- Move maximum to tail --------
    if(maxNode->next != NULL)
    {
        maxPrev->next = maxNode->next;
        struct Node *temp = head;
        while(temp->next != NULL)
            temp = temp->next;
        temp->next = maxNode;
        maxNode->next = NULL;
    }
    return head;
}

File: 09_Linked Lists/03_Move the Smallest and largest to head and tail of list.java
static Node shiftSmallLarge(Node head) {
    if (head == null || head.next == null) return head;
    
    // Find min and max values
    int min = Integer.MAX_VALUE, max = Integer.MIN_VALUE;
    Node curr = head;
    while (curr != null) {
        if (curr.data < min) min = curr.data;
        if (curr.data > max) max = curr.data;
        curr = curr.next;
    }
    
    Node minNode = null, maxNode = null;
    Node newHead = head, tail = head;
    
    // Build a new list excluding min and max nodes
    Node dummy = new Node(0);  // <-- fix: pass 0
    Node ptr = dummy;
    
    curr = head;
    while (curr != null) {
        if (curr.data == min && minNode == null) {
            minNode = curr;
        } else if (curr.data == max && maxNode == null) {
            maxNode = curr;
        } else {
            ptr.next = curr;
            ptr = ptr.next;
        }
        curr = curr.next;
    }
    ptr.next = null;
    
    // Attach max at end
    ptr.next = maxNode;
    maxNode.next = null;
    
    // Attach min at front
    minNode.next = dummy.next;
    
    return minNode;
}